File: MaxSumProblem.cs
using static System.Console;

namespace DP
{
    public class MaxSumProblem
    {
        private int MaximizeCombinedSum(int num)
        {
            int[] dp = new int[num + 1];

            return MaxCombineSum(num);



            int MaxCombineSum(int n)
            {
                // Solve small sub-problems
                //if (n <= 4) return dp[n] = n;   // If num <= 4, combined sum always = num
                if (n <= 11) return dp[n] = n;   // Optimization: If num <= 11, combined sum always = num


                // Divide
                int part = n / 2;   // Sub-problem 1
                int nBy2 = dp[part] == 0 ? dp[part] = MaxCombineSum(part) : dp[part];

                part = n / 3;   // Sub-problem 2
                int nBy3 = dp[part] == 0 ? dp[part] = MaxCombineSum(part) : dp[part];

                part = n / 4;   // Sub-problem 3
                int nBy4 = dp[part] == 0 ? dp[part] = MaxCombineSum(part) : dp[part];



                // Combine
                int sum = nBy2 + nBy3 + nBy4;   // Combined sum
                switch (sum)
                {
                    case int s when s <= n: return dp[n] = n;   // Ignore sum <= num; instead return num
                    default: return dp[n] = sum;   // sum > num; return sum
                }
            }
        }


        internal static void Work()
        {
            int n = 12;
            //int n = 24;

            int maxCombinedSum = new MaxSumProblem().MaximizeCombinedSum(n);
            WriteLine(maxCombinedSum);
        }
    }
}

File: MaxSumProblem.py
class MaxSumProblem:
    def CalculateMaxSumAfterDivisionsMemo(self, num):
        dp = [-1] * (num + 1)

        def CalculateMaxSumAfterDivisions(n):
            # Solve small sub-problems
            if n <= 11:
                dp[n] = n
                return n

            # Divide
            quotient2 = n // 2
            if dp[quotient2] == -1:
                dp[quotient2] = CalculateMaxSumAfterDivisions(quotient2)

            quotient3 = n // 3
            if dp[quotient3] == -1:
                dp[quotient3] = CalculateMaxSumAfterDivisions(quotient3)

            quotient4 = n // 4
            if dp[quotient4] == -1:
                dp[quotient4] = CalculateMaxSumAfterDivisions(quotient4)

            # Combine
            partsSum = dp[quotient2] + dp[quotient3] + dp[quotient4]
            dp[n] = max(n, partsSum)
            return dp[n]

        return CalculateMaxSumAfterDivisions(num)

    def CalculateMaxSumAfterDivisionsTab(self, num):
        dp = [-1] * (num + 1)

        for n in range(num + 1):
            # Solve small sub-problems
            if n <= 11:
                dp[n] = n
                continue

            # Divide
            partsSum = dp[n // 2] + dp[n // 3] + dp[n // 4]

            # Combine
            dp[n] = max(n, partsSum)

        return dp[num]

    @staticmethod
    def Work():
        n = 12  # 13
        # n = 24  # 27

        maxSum = MaxSumProblem().CalculateMaxSumAfterDivisionsTab(n)
        print(maxSum)
